### ISSUE TITLE
[RFE] Allow querying the version (e.g. NEVRA) of the running composer instance.

### ISSUE THREAD

**Issue by thozza** [osbuild/osbuild-composer#1674]
In some cases (e.g. Koji/Brew deployment of osbuild-composer), it is very important that the produced images are always identical for the same input. Deploying a different (newer) version of osbuild-composer may affect the produced images, although we are very focused on ensuring that this does not happen.

It would be beneficial to have a way to query osbuild-composer via some kind of API to get the currently running version (e.g. NEVRA) of it. This information could be then used to determine if there were any changes in the deployed Image Builder infrastructure.

In addition/alternatively, the version of osbuild-composer and osbuild used  for building the image could be provided as an artefact via the Koji API.

**Comment by msehnout** [comment]
> I totally support this. I think our main identifier should be the git sha for both the distro definitions (for the time being that means osbuild-composer), and the git sha of osbuild in the worker. NEVRA might be more tricky as we don't restrict ourselves to deploying releases.

The git SHA will be tricky as well because we don't always build from Git. I'm not sure about production builds in Koji, but when running mock during the execution of `fedpkg scratch-build --srpm`, it doesn't have any git repository so it fails to `git rev-parse HEAD`. Reporting Git revision when available and NEVRA otherwise might work here because there should always be exactly one build with the same NEVRA.

Combined with https://github.com/osbuild/osbuild-composer/issues/1788, osbuild-composer could return something like:
```
build: devel
revision: abc123...
```
or
```
build: prod
revision: osbuild-composer-33-1.f36.x86_64
```

**Comment by ondrejbudai** [comment]
Yeah, when I originally picked this up, I also stumbled upon the same issue. The proposed solution works for me. @teg wdyt?

**Comment by thozza** [comment]
Hmm, I see the problem... However I don't think that having the git SHA is a complete must. The main idea is IMO to be able to determine if the version of any relevant component changed. RPM NEVRA is IMO as good as git SHA, since the SPEC can be used to determine the exact upstream release of sources which was used, plus any additional changes which were backported. 

I would say that reporting only git SHA with RPM builds could be actually wrong. With downstream patches, the SHA would most probably not change, because the base upstream release tarball used for the build would not change. However in such case, we would like composer to report different version.

If we decide that we really want to include the git SHA in all cases, one option could be to always embed it into the sources tarball when it is created from git and then use the value when building binaries. This would probably not work with the release archives created by Github. But as the last resort we could be creating them with a script and uploading to Github? However in case of a production RPM build, I do believe that the NEVRA should be part of the reported version. Nevertheless the upstream git SHA can be there at least as a reference.

**Comment by msehnout** [comment]
@thozza I had exactly the same thoughts about the issues with tarballs. One additional idea that came to my mind is that we could report git sha and the git repository (upstream vs downstream), but that feels overcomplicated and I'm not sure it is even feasible.

Regarding the custom tarballs, I'd prefer to keep away from them as it makes everything more complicated and fragile (building, interaction with Packit, manual uploads).

The problem with using NEVRA only is that our upstream CI will happily build multiple RPMs with the same NEVRA, so we need the additional Git SHA to actually find the right revision.

**Comment by teg** [comment]
The git SHA v NEVRA issue is tricky. We struggled with this when we first set up Schutzbot. The ideal back then was that there should be a direct mapping between the upstream git commit and the downstream RPM (so no downstream patches), and that this should be reflected in the NEVRA. This would allow someone to figure out exactly what code is running for a given NEVRA. This is done for RPMs we build in Schutzbot, but not downstream (yet?).

For this particular use-case, we should keep in mind that the service will eventually only be run from containers, and as such won't have a NEVRA at all (no RPMs are built), so we should at least support exposing the git SHA and only fall back to the NEVRA when no git SHA is set.

Ideally we would figure out a way to always set the git sha in the specfile that is used for our downstream releases, so that we could use the git SHA in all cases. I wonder if that is something Packit has support for and that we could easily hook into? If that's not feasible, falling back to NEVRA might be the next best thing.

**Comment by msehnout** [comment]
If we wanted to keep the tarball generated by Github unchanged and if we wanted to have the same spec file in upstream and downstream, I believe we could create a file in downstream git with a reference to the upstream commit. Something like:
```
Source1:   upstream-revision

%build
# load the %{source1} file
```
On the other hand, we already point to a specific release, which in turn points to a specific upstream revision, so it is already possible with one extra step.

**Comment by thozza** [comment]
@teg did you consider the internal Brew deployment when you mentioned the **service**? While I'm aware of the goal to have the _osbuild-composer_ server running in RHOCP, I'm not sure if this includes the Brew workers. My expectation is that the _osbuild-composer-worker_ will be running on the bare-metal or virtualised workers in the future. Or is there a plan to run workers for each architecture also in a container on our workers? If not, I do believe that we will have to continue deploying our software as internally-built RPMs on them.

Once we move the distro definitions to the distribution itself, we will basically have no other option than using RPMs for them and report the definition's NEVRA.

I also expect that we may need to make a specific change (fix) in the future to our internal Brew service, without pulling in the latest upstream `main`. From my point of view we will have to either use a backported downstream patch in such a case, or have a branch in upstream for it and make a release with some reasonable release version. Upstream release versioning may become a very interesting task in this scenario, if you consider that we may have to maintain multiple variants based on a specific upstream release (for RHEL minor releases, various deployments, ...). It may actually be easier to maintain a change as a downstream patch until we can rebase to some newer upstream release. I'm not advocating that we should default to downstream patches or use them extensively, but I would not disregard them as never happening.

I think that we won't be able to solve this by using only upstream git SHA or RPM NEVRA. It will most probably be a combination of both, when one of them (or both) is used depending on the deployment and build environment. It would be great if we could use the same approach for the whole stack of project which makes up our deployment.

**Comment by msehnout** [comment]
Yes, having a branch for all of our deployments (koji, brew, minor rhel releases) is probably not feasible if we wan't to keep the versioning we have. But I think that is for another discussion.

I think having both Git SHA and NEVRA makes most sense: https://github.com/osbuild/osbuild-composer/pull/1787/files
The users can compare both or just one if the other is undefined.

**Comment by ondrejbudai** [comment]
I think that the version should respect how the RPM was built. For RPMs built by Schutzbot it totally makes sense to use git SHA because Schutzbot is an upstream CI/CD tool and thus it should reflect the upstream versioning.

For RPMs built by koji/brew, it makes sense to use the version that koji/brew uses and that's NEVRA. As Tomas previously said, this also reflects any downstream patches. We're trying to limit their usage but sometimes we just go with them (for various reasons). NEVRA is (almost) unique, so I'm fine with that.

Speaking of almost unique NEVRAs. IIRC, NEVRAs in RHEL and CS can be the same. As these distributions are meant to be binary compatible (whatever that means), it makes sense. Sadly, it isn't as easy for us because we ship different repositories to RHEL than to CS. I wonder if we can also encode inside the RPM if it was built for RHEL or for CS. I know that this is a niche corner case but we never know what users are doing with our software.

@teg About not building RPMs when building a container: This is a pretty large change in our philosophy (everything is an RPM) so I think that changes will be required all over our build pipeline. I, therefore, propose not to worry about it right now and think about it when this time comes. Also, we can just always build RPMs, I don't see a reason why not to.

**Comment by msehnout** [comment]
It seems like we more-less agree that Git SHA is the right way for upstream, but tricky in downstream because we cannot easily make sure there won't be any downstream changes.

I propose we include both for the time being and deal with the remaining questions later. This will unblock the work on how to actually provide the information to users in the brew deployment.

**Comment by thozza** [comment]
Personally I think that not including the Git SHA in downstream build is not a big issue, since the NEVRA defines clearly SRPM used to build binary RPMs (SRPM is part of the build in Brew/Koji) and state of dist-git, from which the RPM was built. Dist-git history defines which sources tarball was used to build the RPM. We should not need worry about random downstream local builds or scratch builds. 

As for distinguishing CentOS and RHEL - although CentOS and RHEL use the same dist tag `elX`, the CentOS packages are in majority of cases just rebuild of the same RHEL version. Both distributions use different signing key and Vendor, so there are additional metadata to distinguish them:
```
[root@ci-vm-10-0-137-231 ~]# rpm -q --info filesystem
Name        : filesystem
Version     : 3.8
Release     : 3.el8
Architecture: x86_64
Install Date: Wed 12 May 2021 06:37:31 AM EDT
Group       : System Environment/Base
Size        : 0
License     : Public Domain
Signature   : RSA/SHA256, Tue 03 Nov 2020 11:03:36 AM EST, Key ID 05b555b38483c65d
Source RPM  : filesystem-3.8-3.el8.src.rpm
Build Date  : Tue 03 Nov 2020 10:23:02 AM EST
Build Host  : x86-02.mbox.centos.org
Relocations : (not relocatable)
Packager    : CentOS Buildsys <bugs@centos.org>
Vendor      : CentOS
URL         : https://pagure.io/filesystem
Summary     : The basic directory layout for a Linux system
Description :
The filesystem package is one of the basic packages that is installed
on a Linux system. Filesystem contains the basic directory layout
for a Linux operating system, including the correct permissions for
the directories.
```
```
[root@ci-vm-10-0-138-5 ~]# rpm -q --info filesystem
Name        : filesystem
Version     : 3.8
Release     : 6.el8
Architecture: x86_64
Install Date: Thu 16 Sep 2021 10:53:53 AM EDT
Group       : System Environment/Base
Size        : 0
License     : Public Domain
Signature   : RSA/SHA256, Mon 21 Jun 2021 07:17:44 AM EDT, Key ID 199e2f91fd431d51
Source RPM  : filesystem-3.8-6.el8.src.rpm
Build Date  : Mon 21 Jun 2021 05:53:08 AM EDT
Build Host  : x86-039.build.eng.bos.redhat.com
Relocations : (not relocatable)
Packager    : Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>
Vendor      : Red Hat, Inc.
URL         : https://pagure.io/filesystem
Summary     : The basic directory layout for a Linux system
Description :
The filesystem package is one of the basic packages that is installed
on a Linux system. Filesystem contains the basic directory layout
for a Linux operating system, including the correct permissions for
the directories.
```

Nevertheless if you'll find a reasonable way to embed the upstream Git SHA into the downstream builds, then it is definitely not a problem to use it.

**Comment by ondrejbudai** [comment]
Hmm, do you see value in including Vendor inside osbuild-composer and its status APIs as well? I'm leaning slightly towards including it... I think I would like to differentiate all of these RHEL rebuilds.

**Comment by thozza** [comment]
I don't see any problem with including the vendor in the version. The value in built RPM is most probably coming from some RPM macro. I noticed that there is `%vendor`, but seems to return `redhat` on Fedora, RHEL and CS. Maybe the build system has it defined to some other value for production builds. In the upstream SPEC, we could define it to `upstream` if not defined...

**Comment by msehnout** [comment]
Just a summary from our internal discussion: It is clear that this feature is not going to solve their use case. I'll focus on fixing this bug: https://github.com/osbuild/osbuild-composer/issues/1788 , but for now there doesn't seem to be a reason to introduce the endpoint.
